File: data/weather.go
package data

import (
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
)

type WeatherAPI struct {
	RootUrl string
	ApiKey  string
}

var Weather = &WeatherAPI{
	RootUrl: "https://api.weatherapi.com/v1",
	ApiKey:  "3824de4dda654325aea05430232312",
}

type WeatherCondition struct {
	Text string  `json:"text"`
	Icon string  `json:"icon"`
	Code float32 `json:"code"`
}

type CurrentWeatherUpstreamResponse struct {
	FeelsLikeC float32          `json:"feelslike_c"`
	FeelsLikeF float32          `json:"feelslike_f"`
	Condition  WeatherCondition `json:"condition"`
	Cloud      float32          `json:"cloud"`
	GustMPH    float32          `json:"gust_mph"`
	GustKPH    float32          `json:"gust_kph"`
	Humidity   float32          `json:"humidity"`
	PrecipIn   float32          `json:"precip_in"`
	PrecipMM   float32          `json:"precip_mm"`
	PressureIn float32          `json:"pressure_in"`
	PressureMb float32          `json:"pressure_mb"`
	TempC      float32          `json:"temp_c"`
	TempF      float32          `json:"temp_f"`
	UV         float32          `json:"uv"`
	VisKM      float32          `json:"vis_km"`
	VisMiles   float32          `json:"vis_miles"`
	WindDegree float32          `json:"wind_degree"`
	WindKPH    float32          `json:"wind_kph"`
	WindMPH    float32          `json:"wind_mph"`
	WindDir    string           `json:"wind_dir"`
}

type AstroCondition struct {
	Sunrise      string  `json:"sunrise"`
	Sunset       string  `json:"sunset"`
	IsSunUp      float32 `json:"is_sun_up"`
	Moonrise     string  `json:"moonrise"`
	Moonset      string  `json:"moonset"`
	MoonPhase    string  `json:"moon_phase"`
	Illumination float32 `json:"moon_illumination"`
}

type DayWeatherUpstreamResponse struct {
	Condition     WeatherCondition `json:"condition"`
	AvgHumidity   float32          `json:"avghumidity"`
	AvgTempC      float32          `json:"avgtemp_c"`
	AvgTempF      float32          `json:"avgtemp_f"`
	AvgVisKM      float32          `json:"avgvis_km"`
	AvgVisMiles   float32          `json:"avgvis_miles"`
	MinTempC      float32          `json:"mintemp_c"`
	MinTempF      float32          `json:"mintemp_f"`
	MaxTempC      float32          `json:"maxtemp_c"`
	MaxTempF      float32          `json:"maxtemp_f"`
	WillItSnow    float32          `json:"daily_will_it_snow"`
	WillItRain    float32          `json:"daily_will_it_rain"`
	WindChillC    float32          `json:"windchill_c"`
	ChanceOfRain  float32          `json:"daily_chance_of_rain"`
	ChanceOfSnow  float32          `json:"daily_chance_of_snow"`
	TotalPrecipIn float32          `json:"totalprecip_in"`
	TotalPrecipMM float32          `json:"totalprecip_mm"`
	TotalSnowCM   float32          `json:"totalsnow_cm"`
	UV            float32          `json:"uv"`
}

type HourWeatherUpstreamResponse struct {
	TimeEpoch int    `json:"time_epoch"`
	Time      string `json:"time"`

	TempC      float32 `json:"temp_c"`
	TempF      float32 `json:"temp_f"`
	WindChillC float32 `json:"windchill_c"`
	WindChillF float32 `json:"windchill_f"`

	Condition WeatherCondition `json:"condition"`
}

type ForecastUpstreamResponse struct {
	Astro AstroCondition                `json:"astro"`
	Day   DayWeatherUpstreamResponse    `json:"day"`
	Hour  []HourWeatherUpstreamResponse `json:"hour"`
}

type ForecastQueryUpstreamResponse struct {
	Current  CurrentWeatherUpstreamResponse `json:"current"`
	Forecast struct {
		ForecastDay []ForecastUpstreamResponse `json:"forecastday"`
	} `json:"forecast"`
}

func (weather *WeatherAPI) TodaysForecast(query string) (*ForecastQueryUpstreamResponse, error) {
	url := fmt.Sprintf(
		"%s/forecast.json?q=%s&key=%s",
		weather.RootUrl,
		query,
		weather.ApiKey,
	)

	resp, err := http.Get(url)
	if err != nil {
		fmt.Println("No response from request")
	}

	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body) // response body is []byte

	if err != nil {
		return &ForecastQueryUpstreamResponse{}, err
	}

	response := ForecastQueryUpstreamResponse{}
	slog.Info("Help", slog.String("body", string(body)))
	err = json.Unmarshal(body, &response)

	if err != nil {
		return &ForecastQueryUpstreamResponse{}, err
	}

	return &response, nil
}

File: client/src/services/weather.ts
import { CurrentWeatherDTO, TodaysWeatherDTO } from './weather.schema'

const ROOT_URL = 'http://localhost:8080/api/v1'

export const transformCurrentWeather = (current: any) => ({
  feelslike: {
    c: current.feelslike_c as number,
    f: current.feelslike_f as number,
  },
  condition: {
    text: current.condition.text,
    icon: current.condition.icon,
    code: current.condition.code,
  },
  cloud: current.cloud as number,
  gust: {
    mph: current.gust_mpg as number,
    kph: current.gust_kph as number,
  },
  humidity: current.humidity as number,
  last_updated: {
    date: current.last_updated as string,
    epoch: current.last_updated_epoch as number,
  },
  precip: {
    in: current.precip_in as number,
    mm: current.precip_mm as number,
  },
  pressure: {
    in: current.pressure_in as number,
    mb: current.pressure_mb as number,
  },
  temp: {
    c: current.temp_c as number,
    f: current.temp_f as number,
  },
  uv: current.uv as number,
  vis: {
    km: current.vis_km as number,
    miles: current.vis_miles as number,
  },
  wind: {
    degree: current.wind_degree as number,
    kph: current.wind_kph as number,
    mph: current.wind_mph as number,
    dir: current.wind_dir as string,
  },
})

export const transformForecastWeather = (forecast: any) => ({
  condition: {
    text: forecast.day.condition.text,
    icon: forecast.day.condition.icon,
    code: forecast.day.condition.code,
  },
  sun: {
    rise: forecast.astro.sunrise,
    set: forecast.astro.sunset,
    is_up: Boolean(forecast.astro.is_sun_up),
  },
  moon: {
    is_up: Boolean(forecast.astro.is_moon_up),
    rise: forecast.astro.moonrise,
    set: forecast.astro.moonset,
    phase: forecast.astro.moon_phase,
    illumination: forecast.astro.moon_illumination,
  },
  avg: {
    humidty: forecast.day.avghumidity,
    temp: {
      c: forecast.day.avgtemp_c,
      f: forecast.day.avgtemp_f,
    },
    vis: {
      km: forecast.day.avgvis_km,
      miles: forecast.day.avgvis_miles,
    },
  },
  min: {
    temp: {
      c: forecast.day.mintemp_c,
      f: forecast.day.mintemp_f,
    },
  },
  max: {
    temp: {
      c: forecast.day.maxtemp_c,
      f: forecast.day.maxtemp_f,
    },
    wind: {
      mph: forecast.day.maxwind_mph,
      kph: forecast.day.maxwind_kph,
    },
  },
  will: {
    rain: Boolean(forecast.day.daily_will_it_rain),
    snow: Boolean(forecast.day.daily_will_it_snow),
  },
  chance: {
    rain: forecast.day.daily_chance_of_rain,
    snow: forecast.day.daily_chance_of_snow,
  },
  total: {
    precip: {
      in: forecast.day.totalprecip_in,
      mm: forecast.day.totalprecip_mm,
    },
    snow: {
      cm: forecast.day.totalsnow_cm,
    },
  },
  uv: forecast.day.uv,
})

export const getCurrentWeather = async (
  query: string,
  token: string,
): Promise<{ current: CurrentWeatherDTO; today: TodaysWeatherDTO }> => {
  const result = await fetch(
    `${ROOT_URL}/weather?query=${encodeURIComponent(query)}`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    },
  )

  return result.json().then((data: any) => {
    const current: CurrentWeatherDTO = transformCurrentWeather(data.current)
    const forecast = data.forecast.forecastday[0]
    const today: TodaysWeatherDTO = transformForecastWeather(forecast)

    return {
      today,
      current,
    }
  })
}
